src/main/java/com/gooddata/warehouse/WarehouseService.java

```diff
@@ -513,35 +515,112 @@ public WarehouseS3Credentials getWarehouseS3Credentials(final Warehouse warehous
-                                            WarehouseS3Credentials.class);
-                            setResult(created);
-                        } catch (GoodDataException | RestClientException e) {
-                            throw new GoodDataException(
-                                    "S3 credentials added to warehouse, but can't get them back, uri: "
-                                            + pollResult.getWarehouseUserUri(), e);
-                        }
-                    }
-
-                    @Override
-                    public void handlePollException(final GoodDataRestException e) {
-                        throw new GoodDataException("Unable to add S3 credentials to warehouse, uri: "
-                                + warehouse.getUri(), e);
-                    }
-                });
+        return new PollResult<>(this, getS3PollHandler(warehouse, s3Credentials, task, "add"));
+    }
+
+
+    /**
+     * update S3 credentials in the Warehouse
+     *
+     * @param warehouse     warehouse to update credentials in
+     * @param s3Credentials the credentials to update
+     * @return updated credentials (not null)
+     */
+    public FutureResult<WarehouseS3Credentials> updateS3CredentialsToWarehouse(final Warehouse warehouse,
+                                                                               final WarehouseS3Credentials s3Credentials) {
+        notNull(warehouse, "warehouse");
+        notEmpty(warehouse.getId(), "warehouse.id");
+        notNull(s3Credentials, "s3Credentials");
+
+        final HttpEntity<WarehouseTask> taskHttpEntity;
+        try {
+            taskHttpEntity = restTemplate.exchange(WarehouseS3Credentials.URI, HttpMethod.PUT,
+                    new HttpEntity<>(s3Credentials), WarehouseTask.class,
+                    warehouse.getId(), s3Credentials.getRegion(), s3Credentials.getAccessKey());
+        } catch (GoodDataException | RestClientException e) {
+            throw new GoodDataException(format("Unable to update S3 credentials in warehouse '%s', region '%s', accessKey '%s'",
+                    warehouse.getId(), s3Credentials.getRegion(), s3Credentials.getAccessKey()), e);
+        }
+        if (taskHttpEntity == null || taskHttpEntity.getBody() == null) {
+            throw new GoodDataException("Empty response when S3 credentials were updated via PUT to API");
+        }
+
+        final WarehouseTask task = taskHttpEntity.getBody();
+
+        return new PollResult<>(this, getS3PollHandler(warehouse, s3Credentials, task, "update"));
+    }
+
+    /**
+     * delete S3 credentials in the Warehouse
+     *
+     * @param warehouse     warehouse to delete credentials in
+     * @param s3Credentials the credentials to delete
+     * @return list of remaining S3 credentials (not null)
+     */
+    public FutureResult<WarehouseS3CredentialsList> deleteS3CredentialsToWarehouse(final Warehouse warehouse,
```

Comment:
why the return type is not void? WarehouseS3CredentialsList doesn't make sense to me